detector/filecontent_detector.go

```diff
@@ -31,66 +31,126 @@ func (fc *FileContentDetector) AggressiveMode() *FileContentDetector {
+	creditCardContent
+)
 
-		hexResults := fc.detectFile(addition.Data, checkHex)
-		fillHexDetectionResults(hexResults, addition, result)
+func (ct contentType) getInfo() string {
+	switch ct {
+	case base64Content:
+		return "Failing file as it contains a base64 encoded text."
+	case hexContent:
+		return "Failing file as it contains a hex encoded text."
+	case creditCardContent:
+		return "Failing file as it contains a potential credit card number."
+	}
+	return ""
+}
 
-		creditCardResults := fc.detectFile(addition.Data, checkCreditCardNumber)
-		fillCreditCardDetectionResults(creditCardResults, addition, result)
+func (ct contentType) getOutput() string {
+	switch ct {
+	case base64Content:
+		return "Expected file to not to contain base64 encoded texts such as: %s"
+	case hexContent:
+		return "Expected file to not to contain hex encoded texts such as: %s"
+	case creditCardContent:
+		return "Expected file to not to contain credit card numbers such as: %s"
 	}
+
+	return ""
 }
 
-func fillResults(results []string, addition gitrepo.Addition, result *DetectionResults, info string, output string) {
-	for _, res := range results {
-		if res != "" {
-			log.WithFields(log.Fields{
-				"filePath": addition.Path,
-			}).Info(info)
+type content struct {
+	name        gitrepo.FileName
+	path        gitrepo.FilePath
+	contentType contentType
+	results     []string
+}
+
+func (fc *FileContentDetector) Test(additions []gitrepo.Addition, ignoreConfig TalismanRCIgnore, result *DetectionResults) {
+	cc := NewChecksumCompare(additions, ignoreConfig)
+	re := regexp.MustCompile(`(?i)checksum[ \t]*:[ \t]*[0-9a-fA-F]+`)
+
+	contents := make(chan content, 512)
+	ignoredFilePaths := make(chan gitrepo.FilePath, len(additions))
+
+	waitGroup := &sync.WaitGroup{}
+	waitGroup.Add(len(additions))
+	for _, addition := range additions {
+		go func(addition gitrepo.Addition) {
+			defer waitGroup.Done()
+			if ignoreConfig.Deny(addition, "filecontent") || cc.IsScanNotRequired(addition) {
+				ignoredFilePaths <- addition.Path
+				return
+			}
+
 			if string(addition.Name) == DefaultRCFileName {
-				result.Warn(addition.Path, "filecontent", fmt.Sprintf(output, res), []string{})
-			} else {
-				result.Fail(addition.Path, "filecontent", fmt.Sprintf(output, res), []string{})
+				content := re.ReplaceAllString(string(addition.Data), "")
+				data := []byte(content)
+				addition.Data = data
+			}
+			contents <- content{
+				name:        addition.Name,
+				path:        addition.Path,
+				contentType: base64Content,
+				results:     fc.detectFile(addition.Data, checkBase64),
+			}
+			contents <- content{
+				name:        addition.Name,
+				path:        addition.Path,
+				contentType: hexContent,
+				results:     fc.detectFile(addition.Data, checkHex),
+			}
+			contents <- content{
+				name:        addition.Name,
+				path:        addition.Path,
+				contentType: creditCardContent,
+				results:     fc.detectFile(addition.Data, checkCreditCardNumber),
 			}
```

Comment:
Lot of repeated code here.
I wonder if repetition could be avoided ?

Reply:
Line 96 to 112, we are just creating 3 structs and sending to channel. I am not very sure how to reduce that. If you have any idea, please let me know.

Reply:
Refactored this. Please review it